### PR TITLE
Raft in Cartridge

### LINKED ISSUE #1233: ☂️ Support native tarantool election_mode as cartridge failover mode
_Modified by @filonenko-mikhail_

## Motivation
There are two failover modes in catridge:
  - eventual
  - stateful
  
Eventual is lightweight but not strong consistent.
Stateful consistent but requires additional sync cross dc kv storage cluster.

Introduce new mode for built-in raft based tarantool failover.

## Bucket Ref On Replica Problem 

Case, when bucket ref was on replica and read the query from replica, lost data during rebalance happened.

- https://github.com/tarantool/vshard/issues/173

## Server disable/enable

The problem when raft switch ro, at this time vshard does not know that request routing changed. vshard router knows about new raft master only after failover timeout + reconfiguration timeout.

If `storage.disable` exists, it happens right after raft switch master flag.

- https://github.com/tarantool/vshard/discussions/307

## We do not know by raft monitoring about leader lost

For e.g. there are three nodes. The first is a leader. And there are two followers.
First two become dead. And the third remember that leader is first. And we can't make the right decision according to the third node world view.

- https://github.com/tarantool/tarantool/issues/6690

## Auto revoke leadership when connectivity loss (r-omk)

- https://github.com/tarantool/tarantool/issues/6661

## (probably) Split-brain when low quorum

- https://github.com/tarantool/tarantool/issues/5295

### PR DESCRIPTION
**Pull request by yngvar-antonsson**
I didn't forget about

- [x] Tests
- [x] Changelog
- [x] Documentation

Close #1233 

Known issues:
- Will not work in case when there are less than 3 nodes in replicaset.
- Could be useful: https://github.com/tarantool/tarantool/commit/e1d2f7f0073cdd893e3bdd032bd48bad33e156bf

TODO:
- Call box.ctl.promote in eventual and stateful mode

### CODE REVIEW

File: cartridge/failover.lua [hunk 1]
```diff
@@ -730,6 +804,57 @@ local function cfg(clusterwide_config)
...
+            -- The instance is set to candidate, so it may become leader itself
+            -- as well as vote for other instances.
+            --
+            -- Alternative: set one of the three instances to `voter`, so that it
+            -- never becomes a leader but still votes for one of its peers and helps
+            -- it reach election quorum (2 in our case).
+            election_mode = os.getenv('TARANTOOL_ELECTION_MODE') or 'candidate',
+            -- Quorum for both synchronous transactions and
+            -- leader election votes.
+            replication_synchro_quorum = failover_cfg.replication_synchro_quorum or 2,
```
Comment (R-omk) [inline cartridge/failover.lua:820]: Since 2.6.2  the option can be expression string
`replication_synchro_quorum = "N/2 + 1"`

File: cartridge/failover.lua [hunk 2]
```diff
@@ -730,6 +804,57 @@ local function cfg(clusterwide_config)
...
+            -- Timeout between elections. Needed to restart elections when no leader
+            -- emerges soon enough.
+            election_timeout = 0.25, -- = 4 * repl_timeout
+        }
+
+        local election = box.info.election
+
+        vars.raft_term = election.term
+
+        local leader = box.info.replication[election.leader] or {}
```
Comment (sergepetrenko) [inline cartridge/failover.lua:837]: Leader won't have time to elect here, so it'll be `{}`.

Is it possible to yield here until leader is found?

File: test/integration/raft_test.lua [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+    end):tomap()
+end
+
+g.before_each(function()
+    h.retrying({}, function()
+        t.assert_equals(h.list_cluster_issues(g.cluster.main_server), {})
+    end)
+    h.retrying({}, function()
+        -- call box.ctl.promote on storage-1
+        set_master('storage-1')
```
Comment (sergepetrenko) [inline test/integration/raft_test.lua:158]: It's not guaranteed that 'storage-1' will be master after calling promote on it.
`box.ctl.promote` runs elections, and it is possible (though highly unlikely), that some other node wins them.
In case the node hasn't won the elections, promote will return an error.

So you may want to do something like `while not ok do ok = pcall(box.ctl.promote) end` to get the expected behaviour.

Or, as an alternative, set everyone except the promoted node to be `follower`.

**Review by sergepetrenko — COMMENTED** [review]
Thanks for the patch!
Please see my comments below.

**Comment by filonenko-mikhail** [comment]
# Related issues:

## Bucket Ref On Replica Problem 

Case when bucket ref was on replica and read query from replica lost data during rebalance happened

- https://github.com/tarantool/vshard/issues/173

File: cartridge/failover.lua [hunk 4]
```diff
@@ -730,6 +735,22 @@ local function cfg(clusterwide_config)
...
+    elseif failover_cfg.mode == 'raft' then
+        raft_failover.cfg(failover_cfg)
+
+        vars.fencing_enabled = false
+        vars.consistency_needed = false
+        first_appointments = raft_failover.get_appointments(topology_cfg)
+
+        vars.failover_fiber = fiber.new(failover_loop, {
+            get_appointments = function()
+                vars.membership_notification:wait()
```
Comment (olegrok) [inline cartridge/failover.lua:747]: Do I correctly understand that any notification will be considered as leader change? Why is it not enough to transfer this info via channel from election trigger? Instances in other replicasets have its own leaders that shouldn't be affected if leader of single replicaset change.

File: cartridge/failover/raft.lua [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    box.cfg{ election_mode = 'off' }
+    vars.raft_term = nil
+    vars.leader_uuid = nil
+end
+
+return {
+    cfg = cfg,
+    disable = disable,
+    get_appointments = get_appointments,
+}
```
Comment (olegrok) [inline cartridge/failover/raft.lua:118]: Manual promotion should be also available. See box.ctl.promote.

File: cartridge/failover/raft.lua [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+    assert(box.ctl.on_election, "Raft failover is available only in Tarantool >= 2.10.0")
+
+    box.cfg{
+        -- The instance is set to candidate, so it may become leader itself
+        -- as well as vote for other instances.
+        --
+        -- Alternative: set one of instances to `voter`, so that it
+        -- never becomes a leader but still votes for one of its peers and helps
+        -- it reach election quorum.
+        election_mode = os.getenv('TARANTOOL_ELECTION_MODE') or 'candidate',
```
Comment (olegrok) [inline cartridge/failover/raft.lua:80]: I guess argparse should be used instead of "os.getenv"

File: test/integration/raft_test.lua [hunk 7]
```diff
@@ -0,0 +1,474 @@
...
+end
+
+g.after_all = function()
+    g.cluster:stop()
+    fio.rmtree(g.cluster.datadir)
+end
+
+local function set_master(instance_name)
+    g.cluster:server(instance_name):exec(function()
+        box.ctl.promote()
```
Comment (olegrok) [inline test/integration/raft_test.lua:127]: I don't see any "box.ctl.promote" call in failover code. Should user call it manually?

File: test/integration/raft_test.lua [hunk 8]
```diff
@@ -0,0 +1,474 @@
...
+        uuid = single_replicaset_uuid,
+    })
+
+    -- send bucket to another storage
+    local bucket_counts = g.cluster:server('storage-1'):exec(function(bucket_id, replicaset_uuid)
+        assert(box.info.ro ~= true)
+        local vshard_storage = require('vshard.storage')
+        vshard_storage.bucket_send(bucket_id, replicaset_uuid)
+        -- wait until sending
+        require('fiber').sleep(0.5)
```
Comment (olegrok) [inline test/integration/raft_test.lua:449]: Why not "wait_for"? Sleeps in code are quite unreliable. Especially on CI where a lot of tasks/containers are run simultaneously.

File: test/integration/raft_test.lua [hunk 9]
```diff
@@ -0,0 +1,474 @@
+local fio = require('fio')
```
Comment (ligurio) [inline test/integration/raft_test.lua:1]: This is a good test suite to start, however, these tests are not enough for testing failover. I would recommend adding  randomness to tests. For example, create a test where we write and read to cluster and periodically kill a leader. Data on cluster must be in a consistent state. Such test for Tarantool was very effective (see https://github.com/tarantool/tarantool/issues/5055#issuecomment-778166416).

File: cartridge/failover.lua [hunk 10]
```diff
@@ -730,6 +735,22 @@ local function cfg(clusterwide_config)
             end,
         })
         vars.failover_fiber:name('cartridge.stateful-failover')
+    elseif failover_cfg.mode == 'raft' then
+        raft_failover.cfg(failover_cfg)
+
+        vars.fencing_enabled = false
+        vars.consistency_needed = false
```
Comment (ligurio) [inline cartridge/failover.lua:742]: Does it mean that we don't need consistency when Raft is enabled?

**Review by ligurio — CHANGES_REQUESTED** [review]

File: doc/schema.graphql [hunk 11]
```diff
@@ -384,6 +256,7 @@ type MutationApicluster {
     """The `Boolean` scalar type represents `true` or `false`."""
     enabled: Boolean
 
+<<<<<<< HEAD
```
Comment (filonenko-mikhail) [inline doc/schema.graphql:259]: dirty merge

File: cartridge/failover.lua (hunk 4 above)
Comment (yngvar-antonsson) [reply cartridge/failover.lua:775]: The current failover mechanism needs to know who is a leader in other replicasets to use this info for rpc calls. Now eventual (and Raft, based on eventual) receives notifications about leader changes and use it to update local leader list.

Comment (yngvar-antonsson) [reply cartridge/failover.lua:775]: The current failover mechanism needs to know who is a leader in other replicasets to use this info for rpc calls. Now eventual (and Raft, based on eventual) receives notifications about leader changes and use it to update local leader list.

Comment (olegrok) [reply cartridge/failover.lua:775]: Any leader change anyway triggers apply_config. I guess topology should be enough.

File: cartridge/failover.lua [hunk 12]
```diff
@@ -753,6 +760,25 @@ local function cfg(clusterwide_config)
             end,
         })
         vars.failover_fiber:name('cartridge.stateful-failover')
+    elseif failover_cfg.mode == 'raft' then
+        local ok, err = ApplyConfigError:pcall(raft_failover.cfg)
+        if not ok then
+            return nil, err
+        end
+
+        vars.fencing_enabled = false
```
Comment (olegrok) [inline cartridge/failover.lua:769]: After https://github.com/tarantool/tarantool/commit/ed46feb27087247ab5ee47fcf63f28c3ad842099 there is `election_fencing_enabled` option. I guess it should be supported.

File: cartridge/failover/raft.lua [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+    local replicasets = assert(topology_cfg.replicasets)
+
+    local appointments = {}
+
+    for replicaset_uuid, _ in pairs(replicasets) do
+        local leaders = topology.get_leaders_order(
+            topology_cfg, replicaset_uuid
+        )
+        if replicaset_uuid == vars.replicaset_uuid then
+            local my_leader_id = box.info.election.leader
```
Comment (olegrok) [inline cartridge/failover/raft.lua:28]: nit: box.info is not so cheap call. Probably it's better to call it once. (it's not serious because it's not hot path).

File: cartridge/failover/raft.lua [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+            local my_leader = box.info.replication[my_leader_id]
+            if my_leader ~= nil then
+                appointments[replicaset_uuid] = my_leader.uuid
+                goto next_rs
+            end
+        end
+
+        local last_leader
```
Comment (olegrok) [inline cartridge/failover/raft.lua:36]: last or latest?

File: cartridge/failover/raft.lua [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+        for _, instance_uuid in ipairs(leaders) do
+            local server = topology_cfg.servers[instance_uuid]
+            local member = membership.get_member(server.uri)
+
+            if member ~= nil
+            and member.payload.leader_uuid ~= nil
+            and member.payload.raft_term >= last_term
+            then
+                last_leader = member.payload.leader_uuid
+                last_term = member.payload.raft_term
```
Comment (olegrok) [inline cartridge/failover/raft.lua:47]: Is it possible that current instance will have "last_term"? Should it be handled anyhow?

Comment (olegrok) [reply cartridge/failover/raft.lua:47]: And if not - why not? I guess comment required.

File: cartridge/failover/raft.lua [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+        end
+        appointments[replicaset_uuid] = last_leader
+        ::next_rs::
+    end
+
+    appointments[vars.replicaset_uuid] = vars.leader_uuid
+    return appointments
+end
+
+local function on_election_trigger()
```
Comment (olegrok) [inline cartridge/failover/raft.lua:58]: I guess this function should trigger failover. Still I'm not sure that failover based on raft should use "eventual" mechanisms. It kills all benefits. Are there any spec/RFC? Who and why suggested it and decided that it should be work so?

File: cartridge/failover/raft.lua [hunk 17]
```diff
@@ -0,0 +1,137 @@
...
+    local election = box.info.election
```
Comment (olegrok) [inline cartridge/failover/raft.lua:59]: nit: the same comment about box.info.

File: cartridge/failover/raft.lua [hunk 18]
```diff
@@ -0,0 +1,137 @@
...
+        -- Quorum for both synchronous transactions and
+        -- leader election votes.
+        replication_synchro_quorum = box_opts.replication_synchro_quorum or 'N/2 + 1',
+        -- Synchronous replication timeout. The transaction will be
+        -- rolled back if no quorum is achieved during timeout.
+        replication_synchro_timeout = box_opts.replication_synchro_timeout,
+        -- Heartbeat timeout. A leader is considered dead if it doesn't
+        -- send heartbeats for 4 * replication_timeout.
+        -- Once the leader is dead, remaining instances start a new election round.
+        replication_timeout = box_opts.replication_timeout,
```
Comment (olegrok) [inline cartridge/failover/raft.lua:93]: This options are not related to raft failover. It's relevant for syncro replication. But it could work with any type failover. That failover just should call "box.ctl.promote" on leader instance.

File: cartridge/failover/raft.lua [hunk 19]
```diff
@@ -0,0 +1,137 @@
...
+    if vars.raft_trigger ~= nil then
+        box.ctl.on_election(nil, vars.raft_trigger)
+        vars.raft_trigger = nil
+    end
+    box.cfg{ election_mode = 'off' }
+    vars.leader_uuid = nil
+end
+
+local function promote(replicaset_leaders)
+    local servers_list = package.loaded.cartridge.confapplier.get_readonly('topology').servers
```
Comment (olegrok) [inline cartridge/failover/raft.lua:119]: I'm not sure but in case of advertise_uri change real value could be different. And "refine_uri..." function should be used.

File: cartridge/failover.lua (hunk 4 above)
Comment (R-omk) [reply cartridge/failover.lua:775]: Suppose there are many instances running on the same host, what happens if the host goes missing?  When for each leader that was on this host, a new one will be elected at the same time, will there not be a race with the loss of actual leader on the other replicasets, or even worse, a  deadloop with a two-phase cfg apply commit?

Everything looks like it need to somehow prevent the massive simultaneous applying the config.

**Review by olegrok — CHANGES_REQUESTED** [review]
This patch doesn't "Close https://github.com/tarantool/cartridge/issues/1372". There is only one mention of box.ctl.promote in your patch.

Remove this "close" from commit messages/PR description or finally implement it. It ~~should~~ must be possible to use synchro with any type failover.

File: cartridge/failover.lua (hunk 4 above)
Comment (olegrok) [reply cartridge/failover.lua:775]: > Everything looks like it need to somehow prevent the massive simultaneous applying the config.

It's relevant for any failover type... And probably it's even implemented.

**Comment by olegrok** [comment]
Following testcase should be covered:
  - Enable raft failover mode
  - Disable raft failover mode (or change it, choose e.g. eventual failover mode) - all raft related box-options should be switched off (in fact mode should be changed to "manual"). E.g. we shouldn't have two failovers working at the same time.

File: cartridge/failover.lua (hunk 4 above)
Comment (yngvar-antonsson) [reply cartridge/failover.lua:775]: As we discussed IRL, leader table not in the topology, so we need membership to notify instances about leader change

**Review by filonenko-mikhail — APPROVED** [review]

**Review by olegrok — CHANGES_REQUESTED** [review]
24 commits can't be pushed as is. Please split patch to atomic changes and add detailed description for each commit.

Patch is too huge and important and should have good history/documentation.

**Review by olegrok — APPROVED** [review]
I left several minor comments below.
